doc/arkode/guide/source/Usage/MRIStep/User_callable.rst

```diff
@@ -171,22 +165,69 @@ MRIStep tolerance specification functions
    This function specifies a user-supplied function *efun* to compute
    the error weight vector ``ewt``.
 
-   **Arguments:**
-      * *arkode_mem* -- pointer to the MRIStep memory block.
-      * *efun* -- the name of the function (of type :c:func:`ARKEwtFn()`)
-        that implements the error weight vector computation.
+   :param arkode_mem: pointer to the MRIStep memory block.
+   :param efun: the name of the function (of type :c:func:`ARKEwtFn()`)
+                that implements the error weight vector computation.
 
-   **Return value:**
-      * *ARK_SUCCESS* if successful
-      * *ARK_MEM_NULL*  if the MRIStep memory was ``NULL``
-      * *ARK_NO_MALLOC*  if the MRIStep memory was not allocated by the time-stepping module
+   :retval ARK_SUCCESS: if successful
+   :retval ARK_MEM_NULL:  if the MRIStep memory was ``NULL``
+   :retval ARK_NO_MALLOC:  if the MRIStep memory was not allocated by the time-stepping module
 
    .. deprecated:: 6.1.0
 
       Use :c:func:`ARKodeWFtolerances` instead.
 
 
 
+.. _ARKODE.Usage.MRIStep.StepAdaptivity:
+
+Multirate temporal adaptivity specification functions
+------------------------------------------------------
+
+As described in Section :numref:`ARKODE.Mathematics.MultirateAdaptivity`, MRIStep
+supports additional forms of temporal adaptivity due to its evolution of multiple
+time scales.
+
+For the simplest form of multirate temporal adaptivity, corresponding to
+"decoupled" multirate control, users should merely specify a single-rate
+controller to MRIStep using the :c:func:`ARKodeSetAdaptController` function.
+The time step controller for the inner integrator may be specified independently.
+
+However, specification of the :math:`h^S-Tol` type of multirate controller requires
+an additional step.  First, a SUNAdaptController with :c:type:`SUNAdaptController_Type` type
+``SUN_ADAPTCONTROLLER_MRI_TOL`` should be constructed, following the documentation in
+Section :numref:`SUNAdaptController.MRIHTol`.  This controller should then be passed as an
+input to the MRIStep function :c:func:`MRIStepSetAdaptController`, as follows.
+
+
+.. c:function:: int MRIStepSetAdaptController(void* arkode_mem, SUNAdaptController C)
```

Comment:
I haven't got to the implementation to understand how difficult this would be, but it would be nice to only need `ARKodeSetAdaptController`. The two function to set the controller take the same arguments, and it seems easy for a user to accidentally call the wrong version.

One idea is to have each stepper in ARKODE (ERK, MRI, etc) optionally provide a function pointer for controller validation/setup to be called in `ARKodeSetAdaptController`. Only MRI would need to specify that, and in that function it could handle special cases with an hS-Tol controller. Again, I'm speculating on a solution since I haven't gotten past the docs.

Reply:
Your idea would work.  Essentially, MRIStep has to create a custom "glue" layer to translate between the hS-Tol controller that it receives, and the normal h-controller that it's ARKODE solver actually calls.

That said, I am reluctant to create an extra layer of indirection for a scenario that would only be used by a single time-stepper.  Perhaps if we wanted to imbue SPRKStep with an adaptivity controller type, where it would similarly translate between a new SUNAdaptController type and the h-controller API that it's own ARKODE solver will call, then this extra layer starts to make sense?  Or something similar for the operator splitting steppers?